### PR TITLE
Add HyperlinkButton

### LINKED ISSUE #7640: Add HyperlinkButton
**Is your feature request related to a problem? Please describe.**

Other XAML frameworks have a [HyperlinkButton](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.hyperlinkbutton) (not to be confused with an inline hyperlink text element inside of a text control). This control is useful to:

 1. Simplify the common use case of a hyperlink button
 2. Provide a standard look-and-feel for hyperlinks in the framework
 3. Provide a cross-platform way of navigating to Uri like webpages.
 4. Automatically manages a change in text color after a link is clicked the first time.

There are additional opportunities here as well. The `HyperlinkButton` could have a new `NavigateUri` property added (like UWP). If a NavigateUri is set, then the HyperlinkButton is pressed, that Uri would be opened in a platform-specific way. This would greatly simplify opening something like a webpage on each platform.

**Describe the solution you'd like**

Add a new HyperlinkButton control derived from button with a standardized style.

**Describe alternatives you've considered**

I've seen often times in the past developers are told to just create their own custom Button style and use that instead.

**Additional context**

### PR DESCRIPTION
**Pull request by robloo**
## What does the pull request do?

Adds a new `HyperlinkButton` control

Depends on: #14320 

## What is the current behavior?

There is no `HyperlinkButton` control

## What is the updated/expected behavior with this PR?

There is now a `HyperlinkButton` control with features beyond those in WinUI.

**Fluent Theme**

image
image

capturegif

**Simple Theme**

image image

## How was the solution implemented (if it's not obvious)?

 1. The UWP/WinUI control was used as a base
 2. Added additional properties discussed in #7640 to control visited/navigation
 3. Used some ideas and code from the previous attempt #9385 

## Checklist

- [ ] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes

None

## Obsoletions / Deprecations

None

## Fixed issues

Closes #7640

### CODE REVIEW

File: src/Avalonia.Controls/HyperlinkButton.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            get => GetValue(IsVisitedProperty);
+            set => SetValue(IsVisitedProperty, value);
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the <see cref="IsVisited"/> property is automatically
+        /// set to true when the <see cref="HyperlinkButton"/> is clicked and the <see cref="NavigateUri"/>
+        /// is successfully launched.
+        /// </summary>
+        public bool IsVisitedOnClick
```
Comment (robloo) [inline src/Avalonia.Controls/HyperlinkButton.cs:61]: The previous attempt used the name `IsVisitedSetOnClick`. I think the `Set` is implied so removed it.

Additionally, there is some debate if the name should be `IsUriVisitedOnClick`. Again, the Uri seems implied but may be better to include it.

File: src/Avalonia.Controls/HyperlinkButton.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        /// Initializes a new instance of the <see cref="HyperlinkButton"/> class.
+        /// </summary>
+        public HyperlinkButton()
+        {
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the <see cref="NavigateUri"/> has been visited.
+        /// </summary>
+        public bool IsVisited
```
Comment (robloo) [inline src/Avalonia.Controls/HyperlinkButton.cs:50]: Open question if this should be `IsUriVisited` to be more clear.

**Comment by robloo** [comment]
cc @rabbitism 

Just keep in mind this is still WIP and isn't complete yet.

File: src/Avalonia.Controls/HyperlinkButton.cs (hunk 1 above)
Comment (thevortexcloud) [reply src/Avalonia.Controls/HyperlinkButton.cs:61]: I am personally in favour of using `IsVisitedSetOnClick`. It sounds far less awkward and is clearer what it is doing. I would also be fine with `SetVisitedOnClick`. That said, I would prefer it not have "click" at all in the name. It conflicts with Avalonia's existing semantics which is pointers.

Comment (robloo) [reply src/Avalonia.Controls/HyperlinkButton.cs]: Property removed and I added comments in NavigateUri for proper usage.

File: src/Avalonia.Themes.Fluent/Controls/FluentControls.xaml [hunk 3]
```diff
@@ -2,6 +2,11 @@
     xmlns="https://github.com/avaloniaui"
     xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
     x:ClassModifier="internal">
+
+  <Style Selector="HyperlinkButton TextBlock">
+    <Setter Property="TextDecorations" Value="Underline"/>
+  </Style>
```
Comment (robloo) [inline src/Avalonia.Themes.Fluent/Controls/FluentControls.xaml:8]: This is not good and I consider it a work-around. However, I'm not sure what else to do here.

Attempting the strategy noted here: https://github.com/AvaloniaUI/Avalonia/issues/7640#issuecomment-1043865206. Doesn't work after control themes. @maxkatz6 Did you come up with a work-around or different strategy for this?

File: src/Avalonia.Controls/HyperlinkButton.cs (hunk 2 above)
Comment (robloo) [reply src/Avalonia.Controls/HyperlinkButton.cs:43]: Probably keeping it short and matching the pseudoclass (as it does currently) is best

File: src/Avalonia.Themes.Fluent/Controls/FluentControls.xaml (hunk 3 above)
Comment (maxkatz6) [reply src/Avalonia.Themes.Fluent/Controls/FluentControls.xaml:8]: Yeah, it's not ideal. But I was going to remove these styles at all, keeping only resource dictionaries.

Comment (maxkatz6) [reply src/Avalonia.Themes.Fluent/Controls/FluentControls.xaml:8]: Will take a look at how we can avoid that

File: src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml [hunk 4]
```diff
@@ -102,6 +102,22 @@
       <StaticResource x:Key="ToggleButtonBorderBrushIndeterminateDisabled"
                       ResourceKey="SystemControlDisabledTransparentBrush" />
 
+      <!-- Resources for HyperlinkButton.xaml -->
+      <StaticResource x:Key="HyperlinkButtonForeground" ResourceKey="SystemControlForegroundBaseHighBrush" />
```
[the remark this answers is not in the document]
Comment (maxkatz6) [reply src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml:106]: Didn't it have accent foreground before? Maybe it was in UWP styles before WinUI.

[the remark this answers is not in the document]
Comment (maxkatz6) [reply src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml:106]: Imo, it made more sense this way.

File: src/Avalonia.Themes.Fluent/Controls/FluentControls.xaml (hunk 3 above)
Comment (maxkatz6) [reply src/Avalonia.Themes.Fluent/Controls/FluentControls.xaml:8]: @robloo TextDecorationsProperty is inherit=true property. You should be able to set `Property="TextBlock.TextDecorations" Value="True"` on HyperlinkButton  styles. I haven't double checked it, but should just work.

Comment (maxkatz6) [reply src/Avalonia.Themes.Fluent/Controls/FluentControls.xaml:8]: > @maxkatz6 Did you come up with a work-around or different strategy for this?

I don't work on that project anymore, so who knows.

Comment (robloo) [reply src/Avalonia.Themes.Fluent/Controls/FluentControls.xaml:8]: > Property="TextBlock.TextDecorations" Value="True"

Yea, just setting the attached property works. I could have sworn I tried that already but guess I made a mistake.

File: src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (robloo) [reply src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml]: I pieced together a new control theme that combines a few ideas together. It is using the accent color now as well.

**Review by maxkatz6 — APPROVED** [review]
LGTM!

**Comment by HEIC-to-JPEG-Dev** [comment]
Is this now live ? I'm can use it in my code but there is no documented api and it's not in the controls list

**Comment by thevortexcloud** [comment]
If it's been merged, it is almost certainly out by now. Docs are generally community contributed (especially for a community made control) and people don't always add them when they add a new control. You can possibly submit a PR to add some here:

https://github.com/AvaloniaUI/avalonia-docs

Or kindly ask Robloo if they can do it.

**Comment by robloo** [comment]
Yes, of course this is available in all Nuget packages now. You can see when the PR was merged. It went out in 11.1 I believe. 

Keep in mind many developers don't have a lot of time. Especially to update documentation for a very simple control like this.

However, the source code and API are fully documented. I made sure to explain in each property what its for and how to use it. You have full contextual documentation in an IDE when you are using the APIs there. You have access to the source code.

You can also go see the UWP/WinUI documentation if you have general questions on why you would use this type of control.

I will push back a little at comments like this because I know you didn't do any legwork yourself before asking.

**Pull request merged**